zebra-consensus/src/transaction.rs

```diff
@@ -224,9 +224,28 @@ impl Request {
 
     /// Returns true if the request is a mempool request.
     pub fn is_mempool(&self) -> bool {
+        matches!(self, Request::Mempool { .. })
+    }
+
+    /// Returns lowest height at which all transparent coinbase spends will be valid,
+    /// or None if this transaction has no transparent coinbase spends.
+    pub fn maturity_height(
+        &self,
+        spent_utxos: &HashMap<transparent::OutPoint, transparent::Utxo>,
+    ) -> Option<block::Height> {
         match self {
-            Request::Block { .. } => false,
-            Request::Mempool { .. } => true,
+            // TODO: return an error for Request::Block to replace this check in the state (#2336)
+            Request::Block { .. } => None,
```

Comment:
In the non-finalized state, mined transactions are verified in the context of their specific chain fork. So this check actually belongs in the state.

(If we try to do it here against the best chain tip, we'll be rejecting some valid blocks on side chains, which stops those side chains later becoming the best chain. This could cause a consensus split.)

Sorry if I was the one to put that note in the ticket, I can see it's wrong now.

Reply:
> (If we try to do it here against the best chain tip, we'll be rejecting some valid blocks on side chains, which stops those side chains later becoming the best chain. This could cause a consensus split.)

It's comparing block (coinbase) heights to spent utxo heights (which should be reliable after #2336), and the transaction verifier calls the state with `AwaitUtxo`  requests to for block transactions so it should find spent UTXOs in any chain.

> Sorry if I was the one to put that note in the ticket, I can see it's wrong now.

I added the possible design, hoping the transaction verifier already had the information to check [`utxo::transparent_spend()`](https://github.com/ZcashFoundation/zebra/blob/2a48d4cf259f82b050e6d08a943844f7d78ef662/zebra-state/src/service/check/utxo.rs) in parallel. I still think that should work after #2336?

Reply:
> > (If we try to do it here against the best chain tip, we'll be rejecting some valid blocks on side chains, which stops those side chains later becoming the best chain. This could cause a consensus split.)
> 
> It's comparing block (coinbase) heights to spent utxo heights (which should be reliable after #2336), and the transaction verifier calls the state with `AwaitUtxo` requests to for block transactions so it should find spent UTXOs in any chain.

I'm not sure, I'd have to think about it a bit more.

It seems to go against the general design principle of doing contextual verification in the state. But that's not a blocker if the change is required to fix a serious bug.

Just making sure we're both on the same page, #2336 wasn't implemented, and it isn't planned any time soon.

Reply:
> It seems to go against the general design principle of doing contextual verification in the state. But that's not a blocker if the change is required to fix a serious bug.

It's not required to fix anything but the transaction verifier seems to already have all of the information needed to do the check in parallel and avoid having the check in both the transaction verifier and the state.

> Just making sure we're both on the same page, https://github.com/ZcashFoundation/zebra/issues/2336 wasn't implemented, and it isn't planned any time soon.

It's currently only a TODO for when #2336 is implemented, it skips the check for block requests.